crawl-ref/source/dungeon.cc

```diff
@@ -6129,6 +6130,132 @@ static bool _valid_item_for_shop(int item_index, shop_type shop_type_,
+ * increasing chance to reroll such items the further down we go.
+ */
+static bool _item_is_trash(item_def& item)
+{
+    // Never trash! (Maybe we could apply *some* logic to random artefacts, but
+    // it's really hard to say what may or may not be useful to a given player.
+    if (is_artefact(item))
+        return false;
+    switch (item.base_type)
+    {
+    case OBJ_SCROLLS:
+        switch (item.sub_type)
+        {
+        case SCR_NOISE:
+        case SCR_IDENTIFY:
+            return true;
+        default:
+            return false;
+        }
+    case OBJ_POTIONS:
+        switch (item.sub_type)
+        {
+        case POT_MOONSHINE:
+        case POT_LIGNIFY:
+            return true;
+        default:
+            return false;
+        }
+    case OBJ_WANDS:
+        switch (item.sub_type)
+        {
+        case WAND_FLAME:
+        case WAND_DIGGING: // Questionable
+            return true;
+        default:
+            return false;
+        }
+    // Get a bit more choosy. A lot of base jewellery types are boring once
+    // you've seen one, but ones with variable plusses can still be good.
+    case OBJ_JEWELLERY:
+        switch (item.sub_type)
+        {
+        case RING_SLAYING:
+        case AMU_FAITH:
+        case AMU_REGENERATION:
+        case AMU_MANA_REGENERATION:
+            return false;
+        case RING_STRENGTH:
+        case RING_INTELLIGENCE:
+        case RING_DEXTERITY:
+        case RING_PROTECTION:
+        case RING_EVASION:
+        case RING_MAGICAL_POWER:
+            return item.plus < 5;
+        default:
+            // All other rings and amulets
+            return true;
+        }
+
+    case OBJ_WEAPONS:
+    {
+        const int rarity = weapon_rarity(item.sub_type);
+        // Rarest weapons are great whatever the stats. Uncommon weapons need
+        // to have a good plus or a brand.
+        return rarity > 6 || rarity > 3 && item.brand == SPWPN_NORMAL && item.plus < 5;
+    }
+
+    case OBJ_MISSILES:
+        // Get rid of the most boring missiles
+        switch (item.sub_type)
+        {
+        case MI_STONE:
+        case MI_BOOMERANG:
+            return true;
+        case MI_JAVELIN:
+            return item.brand == SPMSL_NORMAL;
+        case MI_DART:
+            return item.brand == SPMSL_POISONED;
+        default:
+            return false;
+        }
+    case OBJ_ARMOUR:
+        switch (item.sub_type)
+        {
+        case ARM_ORB:
+            // Non-artefact orbs.
+            return true;
+        case ARM_BARDING:
+            // Barding is rare enough let's allow it even if mundane
+            return false;
+        default:
+            // For everything else, use a bit of logic.
+            return !armour_is_special(item) && item.brand == SPARM_NORMAL
+                   // Eliminates plain +2 hats or boots, but not +5 plate armour
+                   && item.plus < 5;
+        }
```

Comment:
How does this interact with antique shops? Currently it's very often the case that you see e.g. a "glowing cloak" in a shop and you buy it hoping for an ego and it's just a boring +1 cloak. This will potentially significantly change decision making around purchasing such items late in the game (possibly in a good way).

Reply:
What happens is that (for the very late artifact shops), *most* of the items are glowing something or other ... so yes it becomes more of a gamble, but you have a few items to choose from, so maybe buying 2 or 3 things you'll find something nice (and obviously boring +1 cloaks are mostly evicted).